free5gc-operator/controllers/smfdeployment_controller.go

```diff
@@ -0,0 +1,648 @@
+	var ret string
+
+	// n6CfgSlice := getIntConfigSlice(spec.Interfaces, "N6")
+	// n3CfgSlice := getIntConfigSlice(spec.Interfaces, "N3")
+	n4CfgSlice := getIntConfigSlice(spec.Interfaces, "n4")
+	// n9CfgSlice := getIntConfigSlice(spec.Interfaces, "N9")
+
+	ret = `[`
+	intfMap := map[string][]workloadv1alpha1.InterfaceConfig{
+		// "n3": n3CfgSlice,
+		"n4": n4CfgSlice,
+		// "n6": n6CfgSlice,
+		// "n9": n9CfgSlice,
+	}
+	// Need to sort inftMap by key otherwise unitTests might fail as order of intefaces in intfMap is not guaranteed
+	inftMapKeys := make([]string, 0, len(intfMap))
+	for interfaceName := range intfMap {
+		inftMapKeys = append(inftMapKeys, interfaceName)
+	}
+	sort.Strings(inftMapKeys)
+
+	noComma := true
+	for _, key := range inftMapKeys {
+		for _, intf := range intfMap[key] {
+			newNad := fmt.Sprintf(`
+        {"name": "%s",
+         "interface": "%s",
+         "ips": ["%s"],
+         "gateways": ["%s"]
+        }`, constructSMFNadName(templateName, key), intf.Name, intf.IPv4.Address, *intf.IPv4.Gateway)
+			if noComma {
+				ret = ret + newNad
+				noComma = false
+			} else {
+				ret = ret + "," + newNad
+			}
+		}
+	}
+	ret = ret + `
+    ]`
+	return ret
+
+	// return `[
+	//     {"name": "free5gc-smf-1-n4",
+	//      "interface": "n4",
+	//     },
+	// ]`
+
+}
+
+// checkNADExists gets deployment object and checks "k8s.v1.cni.cncf.io/networks" NADs.
+// returns True if all requred NADs are present
+// returns False if any NAD doesn't exists in deployment namespace
+func (r *SMFDeploymentReconciler) checkSMFNADexist(log logr.Logger, ctx context.Context, deployment *appsv1.Deployment) bool {
+	smfAnnotations := []SMFAnnotation{}
+	annotationsString, ok := deployment.Spec.Template.GetAnnotations()["k8s.v1.cni.cncf.io/networks"]
+	if !ok {
+		log.Info("Annotations k8s.v1.cni.cncf.io/networks not found", "SMFDeployment.namespace", deployment.Namespace)
+		return false
+	}
+	if err := json.Unmarshal([]byte(annotationsString), &smfAnnotations); err != nil {
+		log.Info("Failed to parse SMFDeployment annotations", "SMFDeployment.namespace", deployment.Namespace)
+		return false
+	}
+	for _, smfAnnotation := range smfAnnotations {
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(schema.GroupVersionKind{
+			Group:   "k8s.cni.cncf.io",
+			Kind:    "NetworkAttachmentDefinition",
+			Version: "v1",
+		})
+		key := client.ObjectKey{Namespace: deployment.ObjectMeta.Namespace, Name: smfAnnotation.Name}
+		if err := r.Get(ctx, key, u); err != nil {
+			log.Info(fmt.Sprintf("Failed to get NAD %s", smfAnnotation.Name), "SMFDeployment.namespace", deployment.Namespace)
+			return false
+		}
+	}
+
+	return true
+}
+
+func free5gcSMFDeployment(log logr.Logger, smfDeploy *workloadv1alpha1.SMFDeployment) (*appsv1.Deployment, error) {
+	//TODO(jbelamaric): Update to use ImageConfig spec.ImagePaths["smf"],
+	// smfImage := "nephio/free5gc-smf:latest"
+	smfImage := "towards5gs/free5gc-smf:v3.2.0"
```

Comment:
We probably need to be able to read in, at the least, the registry and the tag. But I think this is OK for now (though we have our own built images in the nephio docker now).

Reply:
> We probably need to be able to read in, at the least, the registry and the tag. But I think this is OK for now (though we have our own built images in the nephio docker now).
Agreed, @johnbelamaric , for now it's hardcoded but will add code for reading in the image/version.  As for the docker.io/nephio. I tried referencing that but didn't have access.  Working thru that with @s3wong and SIG Release now.